src/ape/managers/chain.py

```diff
@@ -359,6 +363,113 @@ def revert_to_block(self, block_number: int):
+        Otherwise, caches the contract type to disc at path
+        ``.ape/{ecosystem_name}/{network_name}/contract_types/{address}.json``.
+
+        Args:
+            address (AddressType): The on-chain address of the contract.
+            contract_type (ContractType): The contract's type.
+        """
+
+        if self.get(address):
+            return  # Already cached
+
+        is_local = self._network.name == LOCAL_NETWORK_NAME or self._network.name.endswith("-fork")
+        if is_local and address not in self._local_contracts:
+            self._local_contracts[address] = contract_type
+        else:
+            self._cache_contract_to_disk(address, contract_type)
+
+    def __getitem__(self, address: AddressType) -> ContractType:
+        contract_type = self.get(address)
+        if not contract_type:
+            raise IndexError(f"No contract type found at address '{address}'.")
+
+        return contract_type
+
+    def get(self, address: AddressType) -> Optional[ContractType]:
+        """
+        Get a contract type by address.
+        If the contract is cached, it will return the contract from the cache.
+        Otherwise, if on a live network, it fetches it from the
+        :class:`~ape.api.explorers.ExplorerAPI`.
+
+        Args:
+            address (AddressType): The address of the contract.
+
+        Returns:
+            Optional[ContractType]: The contract type if it was able to get one,
+              otherwise ``None``.
+        """
+
+        contract_type = None
+        is_local = self._network.name == LOCAL_NETWORK_NAME
+        if is_local or self._network.name.endswith("-fork"):
+            # For fork networks, try the local cache first.
+            contract_type = self._local_contracts.get(address)
+
+        if contract_type or is_local:
+            return contract_type
+
+        return self._get_contract_type_from_disk(address) or self._get_contract_type_from_explorer(
+            address
+        )
+
+    def _get_contract_type_from_disk(self, address: AddressType) -> Optional[ContractType]:
+        if not self._contract_types_path.is_dir():
+            return None
+
+        address_file = self._contract_types_path / f"{address}.json"
+        if not address_file.is_file():
+            return None
+
+        contract_type_data = json.loads(address_file.read_text())
+        return ContractType.parse_obj(contract_type_data)
+
+    def _get_contract_type_from_explorer(self, address: AddressType) -> Optional[ContractType]:
+        if not self._network.explorer:
+            return None
+
+        try:
+            contract_type = self._network.explorer.get_contract_type(address)
+        except Exception as err:
+            logger.error(f"Unable to fetch contract type at '{address}' from explorer.\n{err}")
+            return None
+
+        # Cache contract so faster look-up next time.
+        self._cache_contract_to_disk(address, contract_type)
```

Comment:
I believe `ExplorerAPI.get_contract_type` has a return of `Optional[ContractType]` so should probably check it's not `None` before caching?